### DISCUSSION TITLE
Rtsp features

### DISCUSSION THREAD

**Discussion by pedroSG94** [pedroSG94/RootEncoder#879] [Ideas]
This discussion is for join al features or ideas related to RTSP. I will update the main comment with accepted features in progress

**Comment by anilmaddala** [comment]
It would be great if there was support to send KLV metadata over RTSP.

**Reply by pedroSG94** [reply]
Sorry but add KLV metadata to buffers is not worth because most of players/servers don't support it (I could find that gstreamer support decode it but it is no clear how to do it and I can't find a server with this feature).
If you want send extra data to server use a websocket or similar is the recommended way.

**Reply by anilmaddala** [reply]
[rtsp-simple-server](https://github.com/aler9/rtsp-simple-server) has support for KLV data. Is there an RFC/spec/standard I can follow to impement sending KLV metadata over RTSP?

**Reply by pedroSG94** [reply]
This is the RFC for rtp packets:
https://datatracker.ietf.org/doc/html/draft-ietf-payload-rtp-klv-04
And you have my RTP packets here:
https://github.com/pedroSG94/rtmp-rtsp-stream-client-java/tree/master/rtsp/src/main/java/com/pedro/rtsp/rtp/packets

**Reply by anilmaddala** [reply]
For KLV metdata, does the following SDP body lok correct? 
```
fun createKLVBody(trackKLV: Int): String {
        val payload = RtpConstants.payloadType + RtpConstants.trackKLV
        return "m=application 0 RTP/AVP $payload\r\n" +
                "a=rtpmap$payload klv/${RtpConstants.clockVideoFrequency/10}\r\n" +
                "a=control:streamid=$trackKLV\r\n"
}
```

**Reply by pedroSG94** [reply]
> For KLV metdata, does the following SDP body lok correct?
> 
> ```
> fun createKLVBody(trackKLV: Int): String {
>         val payload = RtpConstants.payloadType + RtpConstants.trackKLV
>         return "m=application 0 RTP/AVP $payload\r\n" +
>                 "a=rtpmap$payload klv/${RtpConstants.clockVideoFrequency/10}\r\n" +
>                 "a=control:streamid=$trackKLV\r\n"
> }
> ```

It seems fine. Anyway, if you have problems with the SDP. I good idea is use a program that can use this feature and check packets using a program like wireshark to copy the structure.

**Comment by jambestwick** [comment]
how to record stream rtsp to mp4 file？

**Reply by pedroSG94** [reply]
Hello,

You have an example to record here:
https://github.com/pedroSG94/rtmp-rtsp-stream-client-java/blob/master/app/src/main/java/com/pedro/rtpstreamer/defaultexample/ExampleRtspActivity.java#L173
You can record and stream at same time
